### DISCUSSION TITLE
Remote control 51515 from a Windows application

### DISCUSSION THREAD

**Discussion by brabebhin** [pybricks/support#577] [Q&A]
Hello,

I have been looking into ways to remove control the 51515 set hub from a Windows application. The Windows application should preferably connect to the hub via bluetooth, read some sensor data, and based on the sensor data, compute what to do next. The possible commands would be to continue to move forward at a computed speed,or stop. 

The Windows application is preferably written in C++ as it needs access to some very low hardware functions on the host machine. If possible I could split the application that connects to the hub from the one that computes the speeds through some RPC. The important part is being able to send commands to the hub and get responses in real time.

Is this scenario possible with pybricks?
If it is not possible with 51515 hub, is it possible with other hubs that are currently available on the market?

Thanks.

**Comment by brabebhin** [comment]
Buried deep into the interwebs I found this

https://github.com/pybricks/support/issues/154

Seems to be what I need.
I will be back if I can't get it working.

**Reply by laurensvalk** [reply]
If that issue covers what you need, here's a link that's a little less buried https://pybricks.com/projects/sets/technic/42099-off-roader/keyboard-remote/

Another promising (ongoing) development is this one: https://github.com/pybricks/pybricks-micropython/pull/80

**Reply by brabebhin** [reply]
Thank you for replay. Following the last link I found pybricks.ble. If I understand the docs correctly, I can use that to communicate with any BLE enabled device. Is that true?

**Reply by dlech** [reply]
No, the Pybricks firmware only implements a few Bluetooth services so it won't work with any arbitrary device.

**Reply by brabebhin** [reply]
That includes Windows OS?

**Reply by dlech** [reply]
Yes, OS doesn't matter. (Well, except Apple broke stuff in macOS 12 so that won't work with pybricks/pybricks-micropython#80, but Windows is still good there)

**Reply by brabebhin** [reply]
Sorry, I am confused. Does pybricks.ble work with Windows, as in windows can discover the device and use its characteristics?

**Reply by dlech** [reply]
Actually the proposed changes in the pull request just uses advertising data only, so the computer doesn't even connect to the hub. The just both send and receive advertising data. And yes, Windows is capable of doing this.

**Reply by brabebhin** [reply]
Hi again, @dlech , thanks for your help so far.

I have been poking around the https://github.com/pybricks/pybricksdev-demo repository, trying to recreate those examples in a Visual Studio python environment, but it seems that the API did not import correctly..

the connections.py file does not seem to contain the BLEPUPConnection 

am I missing something?
I installed it using this command: pip install git+https://github.com/pybricks/pybricks-api@master

Thanks.

**Reply by dlech** [reply]
The `pybricksdev-demo` repository has not been kept up to date and should probably be archived.

The `pybricks-api` repository contains the MicroPython API for code that runs on the hubs. This is useful for code completion in VS Code. But programs using this API won't run on your computer.

The `pybricksdev` library is a separate library that runs on your computer and can be used to communicate with hubs. Programs written with this library only run on your computer and won't run on the hub.

**Reply by brabebhin** [reply]
Hi @dlech 
Thanks again for your help.

I did manage to get a simple script up and running, with controlling the hub from the PC.
Now I need to test this with 2 hubs, but I expect it to be the same.

One last question: how long does it generally take for the "output" property of the hub class to synchronize with the physical hub? Is there a way to speed it up a tiny bit?

For whoever will stumble upon this question, here's the code

```python
from pybricksdev.connections import PybricksHub
from pybricksdev.ble import find_device
from bleak.backends.device import BLEDevice
import asyncio
import time

device : BLEDevice = await find_device()
   hub: PybricksHub = PybricksHub()
   await hub.connect(device)

   await hub.run("path to the script", False, True)

   time.sleep(5)

   await hub.write(to send stuff to hub)
```

**Reply by laurensvalk** [reply]
New lines become available in hub.output when a new line is received.

This would be every time you print a line using print() or sooner if you send the end of line symbol \r\n explicitly.

**Reply by brabebhin** [reply]
Hi @laurensvalk 

Thanks for the prompt response!

You mean using the "print" function in the hub (physical hub)? There still seems to be some delay. Is there another way to read stuff from the hub? Like reading from a GATT characteristics or something?

Thanks.

**Reply by laurensvalk** [reply]
Yes, that's the one.

The class you are using looks like this: https://github.com/pybricks/pybricksdev/blob/master/pybricksdev/connections.py#L150

Perhaps you can modify it to suit what you are trying to do.

**Reply by laurensvalk** [reply]
It looks like you are using a blocking sleep call. This causes everything else to block for 5 seconds. Is this the delay you are seeing? 

You'll want to replace it with

await asyncio.sleep(5)

**Reply by brabebhin** [reply]
You are completely right. Now it seems to update in a timely manner. I completely missed the async stuff.
Thanks!

**Comment by lobodpav** [comment]
Tried the Keyboard remote control and it works fine in https://code.pybricks.com

When trying `pybricksdev`, it doesn't. I recall `pybricksdev` not supporting sending commands back to the hub from some other discussion.

I was wondering, would there be any way to run a program on the Hub and connect it to, say PlayStation gamepad?

Asking because currently when I want BT remote control, I use Lego FW. When I need an autonomous robot, I re-flash Pybricks. This is very inconvenient and I'd like to use `pybricks` only if possible, and ideally without the need for a computer. I.e. the Hub and a remote controller only.

**Reply by dlech** [reply]
> When trying pybricksdev, it doesn't. I recall `pybricksdev` not supporting sending commands back to the hub from some other discussion.

https://github.com/pybricks/pybricksdev/issues/17

> I was wondering, would there be any way to run a program on the Hub and connect it to, say PlayStation gamepad?

https://github.com/pybricks/support/issues/191
